sentry-core/src/metrics.rs

```diff
@@ -0,0 +1,387 @@
+                std::io::ErrorKind::Other,
+                "Flushing Client failed",
+            ))
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+enum MetricType {
+    Counter,
+    Distribution,
+    Set,
+    Gauge,
+}
+
+impl MetricType {
+    /// Return the shortcode for this metric type.
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            MetricType::Counter => "c",
+            MetricType::Distribution => "d",
+            MetricType::Set => "s",
+            MetricType::Gauge => "g",
+        }
+    }
+}
+
+impl fmt::Display for MetricType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(self.as_str())
+    }
+}
+
+impl std::str::FromStr for MetricType {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(match s {
+            "c" | "m" => Self::Counter,
+            "h" | "d" | "ms" => Self::Distribution,
+            "s" => Self::Set,
+            "g" => Self::Gauge,
+            _ => return Err(()),
+        })
+    }
+}
+
+struct GaugeValue {
+    last: f64,
+    min: f64,
+    max: f64,
+    sum: f64,
+    count: u64,
+}
+enum BucketValue {
+    Counter(f64),
+    Distribution(Vec<f64>),
+    Set(BTreeSet<String>),
+    Gauge(GaugeValue),
+}
+impl BucketValue {
+    fn distribution(val: f64) -> BucketValue {
+        Self::Distribution(vec![val])
+    }
+
+    fn gauge(val: f64) -> BucketValue {
+        Self::Gauge(GaugeValue {
+            last: val,
+            min: val,
+            max: val,
+            sum: val,
+            count: 1,
+        })
+    }
+
+    fn set_from_str(value: &str) -> BucketValue {
+        BucketValue::Set([value.into()].into())
+    }
+
+    fn merge(&mut self, other: BucketValue) -> Result<(), ()> {
+        match (self, other) {
+            (BucketValue::Counter(c1), BucketValue::Counter(c2)) => {
+                *c1 += c2;
+            }
+            (BucketValue::Distribution(d1), BucketValue::Distribution(d2)) => {
+                d1.extend(d2);
+            }
+            (BucketValue::Set(s1), BucketValue::Set(s2)) => s1.extend(s2),
+            (BucketValue::Gauge(g1), BucketValue::Gauge(g2)) => {
+                g1.last = g2.last;
+                g1.min = g1.min.min(g2.min);
+                g1.max = g1.max.max(g2.max);
+                g1.sum += g2.sum;
+                g1.count += g2.count;
+            }
+            _ => return Err(()),
+        }
+        Ok(())
+    }
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+struct BucketKey {
+    timestamp: u64,
+    ty: MetricType,
+    name: String,
+    tags: String,
+}
+
+type AggregateMetrics = BTreeMap<BucketKey, BucketValue>;
+
+enum Task {
+    RecordMetric((BucketKey, BucketValue)),
+    Flush,
+    Shutdown,
+}
+
+pub struct MetricFlusher {
+    sender: SyncSender<Task>,
+    handle: Option<JoinHandle<()>>,
+}
+
+const FLUSH_INTERVAL: Duration = Duration::from_secs(10);
```

Comment:
This constant is used for two separate concerns, I'd suggest to separate them: 
 1. The bucketing interval, to which timestamps will be rounded when reporting metrics upstream.
 2. The interval for flushing buckets upstream.

Right now, I do not see a strong reason why those two have to be different, but it is important not to change the bucketing interval once we choose a different flush interval.